ms2deepscore/train_new_model/inchikey_pair_selection.py

```diff
@@ -250,83 +251,110 @@ def select_balanced_pairs(available_pairs_for_bin_matrix: np.ndarray,
-                inchikey_with_lowest_count][pair_indexes_with_min_count]
+            # Get second inchikeys for pairs with minimal pair frequency
+            second_inchikeys = available_pairs_row[min_freq_mask]
 
-            second_inchikey_with_lowest_count = available_inchikeys_with_min_count[np.argmin(
-                inchikey_counts[available_inchikeys_with_min_count])]
+            # Get counts for second inchikeys
+            second_inchikey_counts = inchikey_counts[second_inchikeys]
 
-            # Add the selected pairs to pair_frequency:
-            position_of_second_inchikey_in_matrix = available_pairs_for_bin_matrix[
-                                                        inchikey_with_lowest_count] == second_inchikey_with_lowest_count
-            pair_frequency[inchikey_with_lowest_count][position_of_second_inchikey_in_matrix] += 1
+            # Select the second inchikey with the lowest count
+            min_count_idx = np.argmin(second_inchikey_counts)
+            second_inchikey_with_lowest_count = second_inchikeys[min_count_idx]
 
+            # Update pair frequency
+            pair_indices = np.where(available_pairs_row == second_inchikey_with_lowest_count)[0]
+            pair_frequency[inchikey_with_lowest_count, pair_indices] += 1
+
+            # If the pair is not symmetrical, update the reverse pair frequency
             if second_inchikey_with_lowest_count != inchikey_with_lowest_count:
-                # also increase pair_frequency if duplicate
-                position_of_first_inchikey_in_matrix = available_pairs_for_bin_matrix[
-                                                           second_inchikey_with_lowest_count] == inchikey_with_lowest_count
-                pair_frequency[second_inchikey_with_lowest_count][position_of_first_inchikey_in_matrix] += 1
-            # increase inchikey counts
+                reverse_pairs_row = available_pairs_for_bin_matrix[second_inchikey_with_lowest_count]
+                reverse_pair_indices = np.where(reverse_pairs_row == inchikey_with_lowest_count)[0]
+                pair_frequency[second_inchikey_with_lowest_count, reverse_pair_indices] += 1
+
+            # Update inchikey counts
             inchikey_counts[inchikey_with_lowest_count] += 1
             inchikey_counts[second_inchikey_with_lowest_count] += 1
+
             nr_of_pairs_selected += 1
             progress_bar.update(1)
+
+            # If this inchikey still has valid pairs, push it back into the heap
+            if np.any(pair_frequency[inchikey_with_lowest_count] < max_resampling):
+                heapq.heappush(available_inchikey_indexes,
+                               (inchikey_counts[inchikey_with_lowest_count], inchikey_with_lowest_count))
+
+            # Also check if the second inchikey needs to be added back to the heap
+            if second_inchikey_with_lowest_count != inchikey_with_lowest_count:
```

Comment:
The second inchikey is never popped from the available inchikey indexes, so I believe it should never be added back in.